devito/passes/iet/misc.py

```diff
@@ -77,47 +90,68 @@ def relax_incr_dimensions(iet, **kwargs):
-        body = compose_nodes(outer)
-        dynamic_parameters = flatten((i.symbolic_bounds, i.step) for i in outer)
-        dynamic_parameters.extend([i.step for i in inner if not is_integer(i.step)])
-        efunc = make_efunc(name, body, dynamic_parameters)
-
-        efuncs.append(efunc)
-
-        # Create the ElementalCalls
-        calls = []
-        for p in product(*ranges):
-            dynamic_args_mapper = {}
-            for i, (m, M, b) in zip(outer, p):
-                dynamic_args_mapper[i.symbolic_min] = m
-                dynamic_args_mapper[i.symbolic_max] = M
-                dynamic_args_mapper[i.step] = b
-                for j in inner:
-                    if j.dim.root is i.dim.root and not is_integer(j.step):
-                        value = j.step if b is i.step else b
-                        dynamic_args_mapper[j.step] = (value,)
-            calls.append(efunc.make_call(dynamic_args_mapper))
-
-        mapper[root] = List(body=calls)
-
-    iet = Transformer(mapper).visit(iet)
-
-    return iet, {'efuncs': efuncs}
+        # Get symbolic_max out of each outer dimension
+        roots_max = {i.dim.root: i.symbolic_max for i in outer}
+
+        # A dictionary to map maximum of parent dimensions
+        # useful for hierarchical blocking
+        proc_parents_max = {}
+
+        # Process inner iterations
+        for n, i in enumerate(inner):
+            assert i.direction is Forward
+            if (i.dim.parent in proc_parents_max and
+               i.symbolic_size == i.dim.parent.step):
+                # Special case: A parent dimension may have already been processed
+                # as a member of 'inner' iterations. In this case we can use parent's
+                # Max
+                # Usually encountered in hierarchical blocking (BLOCKLEVELS > 1)
+                it_max = proc_parents_max[i.dim.parent]
+            else:
+                # Most of the cases pass though this code:
+                # Candidates for upper bound calculation are:
+                # Candidate 1: symbolic_max of current iteration
+                # e.g.
+                # i.symbolic_max = x0_blk0 + x0_blk0_size
+                symbolic_max = i.symbolic_max
+
+                # Candidate 2: The domain max. Usualy it is the max of parent/root
+                # dimension.
+                # e.g. x_M
+                # This may not always be true as the symbolic_size of an Iteration may
+                # exceed the size of a parent's block size (e.g. after CIRE passes)
+                # e.g.
+                # i.dim.parent.step = x0_blk1_size
+                # i.symbolic_size = x0_blk1_size + 4
+
+                # For this case, proper margin should be allowed
+                # in order not to drop iterations. So Candidate 2 is the maximum of the
+                # root's max and the current iteration's required max
+                # and instead of `x_M` we may need `x_M + 1` or `x_M + 2`
```

Comment:
when isn't the "Iteration's required max" larger than the root's max?

Reply:
Some weird CIRE opt edge cases, need to dig a bit!

Reply:
Actually in test_dse it is the following cases that need this:
```
FAILED tests/test_dse.py::TestAliases::test_min_storage_in_isolation - bdb.BdbQuit
FAILED tests/test_dse.py::TestAliases::test_full_shape_big_temporaries - bdb.BdbQuit
FAILED tests/test_dse.py::TestAliases::test_ftemps_option - bdb.BdbQuit

```